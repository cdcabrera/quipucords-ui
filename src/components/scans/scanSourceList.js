import React from 'react';
import PropTypes from 'prop-types';
import {
  Alert,
  AlertVariant,
  EmptyState,
  EmptyStateVariant,
  List,
  ListItem,
  ListVariant,
  Spinner
} from '@patternfly/react-core';
import { ContextIcon, ContextIconVariant } from '../contextIcon/contextIcon';
import { useGetScanJob } from './scansContext';
import { translate } from '../i18n/i18n';

/**
 * Return a scan jobs listing for "sources".
 *
 * @param {object} props
 * @param {string} props.id
 * @param {Function} props.t
 * @param {Function} props.useGetScanJob
 * @returns {React.ReactNode}
 */
const ScanSourceList = ({ id, t, useGetScanJob: useAliasGetScanJob }) => {
  const { error, errorMessage, pending, data } = useAliasGetScanJob(id);

  if (pending) {
    return (
      <EmptyState className="quipucords-empty-state" variant={EmptyStateVariant.large}>
        <Spinner isSVG size="sm" /> {t('view.loading')}
      </EmptyState>
    );
  }

  if (error) {
    return (
      <EmptyState className="quipucords-empty-state__alert">
        <Alert isInline isPlain variant={AlertVariant.danger} title={t('view.error', { context: 'scan-jobs' })}>
          {t('view.error-message', { context: ['scan-jobs'], message: errorMessage })}
        </Alert>
      </EmptyState>
    );
  }

  return (
    <List isPlain>
      {data?.map(({ id: sourceId, taskStatusMessage, taskStatus, taskType, name, sourceType }) => (
        <ListItem key={sourceId}>
          <List isPlain variant={ListVariant.inline}>
            <ListItem key={name}>
              <ContextIcon symbol={ContextIconVariant[sourceType]} /> {name}
            </ListItem>
            <ListItem key={`desc-${name}`}>
              {t('table.label', {
                context: ['scan-job', taskType, (taskStatusMessage && 'message') || (taskStatus && 'status')],
                status: taskStatus,
                message: taskStatusMessage
              })}
            </ListItem>
          </List>
        </ListItem>
      ))}
    </List>
  );
};

/**
 * Prop types
 *
 * @type {{t: Function, id: string|number, useGetScanJob: Function}}
 */
ScanSourceList.propTypes = {
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  t: PropTypes.func,
  useGetScanJob: PropTypes.func
};

/**
 * Default props
 *
 * @type {{t: translate, useGetScanJob: Function}}
 */
ScanSourceList.defaultProps = {
  t: translate,
  useGetScanJob
};

export { ScanSourceList as default, ScanSourceList };
