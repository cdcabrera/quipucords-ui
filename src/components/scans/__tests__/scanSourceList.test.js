import React from 'react';
import configureMockStore from 'redux-mock-store';
import { mount, shallow } from 'enzyme';
import { ConnectedScanSourceList, ScanSourceList } from '../scanSourceList';

describe('ScanSourceList Component', () => {
  const generateEmptyStore = (obj = {}) => configureMockStore()(obj);

  it('should render a connected component', () => {
    const store = generateEmptyStore({ scansAction: { job: {} } });

    const props = {
      id: 1
    };

    const component = shallow(<ConnectedScanSourceList {...props} />, {
      context: { store }
    });

    expect(component).toMatchSnapshot('connected');
  });

  it('should render a non-connected component', () => {
    const props = {
      id: 1,
      scanJobList: [
        {
          connectTaskStatus: 'lorem',
          connectTaskStatusMessage: 'ipsum',
          id: 1,
          inspectTaskStatus: 'dolor',
          inspectTaskStatusMessage: 'sit',
          name: 'test',
          sourceType: ''
        }
      ]
    };

    const component = mount(<ScanSourceList {...props} />);

    expect(component.render()).toMatchSnapshot('non-connected');
  });

  it('should render a non-connected component error', () => {
    const props = {
      id: 1,
      error: true,
      errorMessage: 'Lorem Ipsum.'
    };

    const component = shallow(<ScanSourceList {...props} />);
    expect(component).toMatchSnapshot('non-connected error');
  });

  it('should render a non-connected component pending', () => {
    const props = {
      id: 1,
      pending: true
    };

    const component = shallow(<ScanSourceList {...props} />);
    expect(component).toMatchSnapshot('non-connected pending');
  });

  it('should handle multiple status messages', () => {
    const props = {
      id: 1,
      scanJobList: [
        {
          connectTaskStatus: 'pending',
          connectTaskStatusMessage: 'ipsum',
          id: 1,
          inspectTaskStatus: 'dolor',
          inspectTaskStatusMessage: 'sit',
          name: 'test',
          sourceType: ''
        }
      ]
    };

    let component = shallow(<ScanSourceList {...props} />);
    expect(component).toMatchSnapshot('connect status');

    props.scanJobList[0].connectTaskStatus = 'completed';
    component = shallow(<ScanSourceList {...props} />);
    expect(component).toMatchSnapshot('inspect status');

    props.scanJobList[0].inspectTaskStatus = undefined;
    component = shallow(<ScanSourceList {...props} />);
    expect(component).toMatchSnapshot('fallback status');
  });
});
