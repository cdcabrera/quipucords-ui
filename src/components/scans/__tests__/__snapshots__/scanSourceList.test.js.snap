// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScanSourceList Component should handle multiple status messages: connect status 1`] = `
<Grid
  bsClass="container"
  componentClass="div"
  fluid={true}
>
  <Row
    bsClass="row"
    className="fadein"
    componentClass="div"
    key="1"
  >
    <Col
      bsClass="col"
      componentClass="div"
      md={3}
      xs={4}
    >
      <Icon
        name=""
        type="pf"
      />
        
      test
    </Col>
    <Col
      bsClass="col"
      componentClass="div"
      md={9}
      xs={8}
    >
      Connection Scan: ipsum
    </Col>
  </Row>
</Grid>
`;

exports[`ScanSourceList Component should handle multiple status messages: fallback status 1`] = `
<Grid
  bsClass="container"
  componentClass="div"
  fluid={true}
>
  <Row
    bsClass="row"
    className="fadein"
    componentClass="div"
    key="1"
  >
    <Col
      bsClass="col"
      componentClass="div"
      md={3}
      xs={4}
    >
      <Icon
        name=""
        type="pf"
      />
        
      test
    </Col>
    <Col
      bsClass="col"
      componentClass="div"
      md={9}
      xs={8}
    >
      Connection Scan: ipsum
    </Col>
  </Row>
</Grid>
`;

exports[`ScanSourceList Component should handle multiple status messages: inspect status 1`] = `
<Grid
  bsClass="container"
  componentClass="div"
  fluid={true}
>
  <Row
    bsClass="row"
    className="fadein"
    componentClass="div"
    key="1"
  >
    <Col
      bsClass="col"
      componentClass="div"
      md={3}
      xs={4}
    >
      <Icon
        name=""
        type="pf"
      />
        
      test
    </Col>
    <Col
      bsClass="col"
      componentClass="div"
      md={9}
      xs={8}
    >
      Inspection Scan: sit
    </Col>
  </Row>
</Grid>
`;

exports[`ScanSourceList Component should render a connected component: connected 1`] = `
<ScanSourceList
  error={false}
  errorMessage={null}
  getScanJob={[Function]}
  id={1}
  pending={false}
  scanJobList={Array []}
/>
`;

exports[`ScanSourceList Component should render a non-connected component error: error 1`] = `
<EmptyState
  className=""
>
  <EmptyStateIcon
    className=""
    name="error-circle-o"
    type="pf"
  />
  <EmptyStateTitle
    className=""
  >
    Error retrieving scan jobs
  </EmptyStateTitle>
  <EmptyStateInfo
    className=""
  >
    Lorem Ipsum.
  </EmptyStateInfo>
</EmptyState>
`;

exports[`ScanSourceList Component should render a non-connected component pending: pending 1`] = `
<EmptyState
  className=""
>
  <EmptyStateIcon
    className=""
    name="spinner spinner-xl"
    type="pf"
  />
  <EmptyStateTitle
    className=""
  >
    Loading scan jobs...
  </EmptyStateTitle>
</EmptyState>
`;

exports[`ScanSourceList Component should render a non-connected component: non-connected 1`] = `
<div
  class="container-fluid"
>
  <div
    class="fadein row"
  >
    <div
      class="col-md-3 col-xs-4"
    >
      <span
        aria-hidden="true"
        class="pficon pficon-"
      />
        test
    </div>
    <div
      class="col-md-9 col-xs-8"
    >
      Connection Scan: ipsum
    </div>
  </div>
</div>
`;
